doc/subsystems/codegen/modules.rst

```diff
@@ -0,0 +1,316 @@
+    initialisation that can not be accomplished by ``device_declare``.
+
+    :param device_config:
+        Configuration variables for device instantiation.
+        (e.g. 'CONFIG_SPI_0')
+    :param driver_name:
+        Driver name for device instantiation.
+        (e.g. 'SPI_0')
+    :param device_init:
+        Device initialisation function.
+        (e.g. 'spi_stm32_init')
+    :param device_level:
+        Driver initialisation level.
+        (e.g. 'PRE_KERNEL_1')
+    :param device_prios:
+        Driver initialisation priority definition.
+        (e.g. 32)
+    :param device_api:
+        Identifier of the device api.
+        (e.g. 'spi_stm32_driver_api')
+    :param device_info:
+        Device info template for device driver config, data and interrupt
+        initialisation.
+
+Raw declaration of multiple device instances
+--------------------------------------------
+
+.. function:: devicedeclare.device_declare_multi(device_configs, driver_names, device_inits, device_levels, device_prios, device_api, device_info)
+
+    Generate device instances code for all device instances that:
+
+    - match the driver names and that
+    - are activated ('status' = 'ok') in the board device tree file and that are
+    - configured by Kconfig.
+
+    The ``device_declare_multi`` function enables a detailed control of the
+    device info definition. It's primary use is for complex device instance
+    initialisation that can not be accomplished by ``device_declare``.
+
+    :param device_configs:
+        A list of configuration variables for device instantiation.
+        (e.g. ['CONFIG_SPI_0', 'CONFIG_SPI_1'])
+    :param driver_names:
+        A list of driver names for device instantiation. The list shall be
+        ordered the same way as the list of device configs.
+        (e.g. ['SPI_0', 'SPI_1'])
+    :param device_inits:
+        A list of device initialisation functions or a one single function.
+        The list shall be ordered as the list of device configs.
+        (e.g. 'spi_stm32_init')
+    :param device_levels:
+        A list of driver initialisation levels or one single level definition.
+        The list shall be ordered as the list of device configs.
+        (e.g. 'PRE_KERNEL_1')
+    :param device_prios:
+        A list of driver initialisation priorities or one single priority
+        definition. The list shall be ordered as the list of device configs.
+        (e.g. 32)
+    :param device_api:
+        Identifier of the device api.
+        (e.g. 'spi_stm32_driver_api')
+    :param device_info:
+        Device info template for device driver config, data and interrupt
+        initialisation.
+
+Example:
+
+.. code-block:: C
+
+    /**
+     * @code{.codegen}
+     * codegen.import_module('devicedeclare')
+     *
+     * device_configs = ['CONFIG_SPI_{}'.format(x) for x in range(1, 4)]
+     * driver_names = ['SPI_{}'.format(x) for x in range(1, 4)]
```

Comment:
See above. The driver writer can choose a upper limit that covers all future SoCs. As only the configured and activated devices will be intiantiated the upper limit can be any big number.

Reply:
> Besides, couldn't we use dts property 'label' instead of yet another instance name.

The driver name is the same as the 'label' property. In the INIT macro it is called driver_name so I stick with it. The extended DTS database is searched for a matching 'label' property to retrieve the DTS properties of the device instance with the given driver name.